### PR TITLE
add support to load kubernetes configuration for memory buffer

### PR DESCRIPTION
**Pull request by DanyT**
- existing api is not changed
   - load_kube_config works as before
- extend kubeconfig_t to hold in memory buffer
- make a common flow for file/buffer code
- set fileName/buffer members accordingly
- provide new api: load_kube_config_buffer

### CODE REVIEW

**Comment by ityuhui** [comment]
Thanks for your PR. I will review it.

File: kubernetes/config/kube_config.c [hunk 1]
```diff
@@ -393,8 +386,48 @@ int load_kube_config(char **pBasePath, sslConfig_t ** pSslConfig, list_t ** pApi
...
   end:
+    return rc;
+}
+
+int load_kube_config(char **pBasePath, sslConfig_t ** pSslConfig, list_t ** pApiKeys, const char *configFileName)
+{
+    static char fname[] = "load_kube_config()";
+    int rc = 0;
+
+    kubeconfig_t *kubeconfig = kubeconfig_create();
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:397]: Given that you delete this below, I think you can stack allocate this and pass the pointer via `&` below.

That will save a heap allocation and also be cleaner code.

File: kubernetes/config/kube_config.c [hunk 2]
```diff
@@ -393,8 +386,48 @@ int load_kube_config(char **pBasePath, sslConfig_t ** pSslConfig, list_t ** pApi
...
+
+    return rc;
+}
+
+int load_kube_config_buffer(char **pBasePath, sslConfig_t ** pSslConfig, list_t ** pApiKeys, const char *buffer)
+{
+    static char fname[] = "load_kube_config_buffer()";
+    int rc = 0;
+
+    kubeconfig_t *kubeconfig = kubeconfig_create();
```
Comment (brendandburns) [inline kubernetes/config/kube_config.c:418]: Stack allocate here to, as above.

File: kubernetes/config/kube_config_yaml.h [hunk 3]
```diff
@@ -24,13 +24,15 @@ extern "C" {
  *
  * IN:
  * kubeconfig->fileName: kubernetes cluster configuration file name
```
Comment (brendandburns) [inline kubernetes/config/kube_config_yaml.h:26]: Maybe note that either one, but not both should be supplied?

File: kubernetes/config/kube_config_yaml.c [hunk 4]
```diff
@@ -430,25 +430,34 @@ int kubeyaml_load_kubeconfig(kubeconfig_t * kubeconfig)
 {
     static char fname[] = "kubeyaml_load_kubeconfig()";
 
-    /* Set a file input. */
+    /* Set a file input or use the provided buffer. */
     FILE *input = NULL;
```
Comment (brendandburns) [inline kubernetes/config/kube_config_yaml.c:434]: Add a test here to verify that only one of `fileName` and `buffer` is non-null.

**Comment by brendandburns** [comment]
Minor comments.

File: kubernetes/config/kube_config.c (hunk 1 above)
Comment (DanyT) [reply kubernetes/config/kube_config.c:397]: This can be done but implies additional changes in kubeconfig_free() as this function fill attempt to free the kubeconfig_t* itself.  I can do a kubeconfig_free_static() if needed in order to keep the original api interface unchanged.

Please let me know your thoughts on this one.

File: kubernetes/config/kube_config.c (hunk 2 above)
Comment (DanyT) [reply kubernetes/config/kube_config.c:418]: Same here. See https://github.com/kubernetes-client/c/pull/242#discussion_r1663546396

File: kubernetes/config/kube_config_yaml.h (hunk 3 above)
Comment (DanyT) [reply kubernetes/config/kube_config_yaml.h:26]: I have added a more explicit note there. There was already a hint in the doc for **kubeconfig->buffer**.
`* IN:
 * kubeconfig->fileName: kubernetes cluster configuration file name
 * kubeconfig->buffer: kubernetes cluster configuration data; **this is considered only if kubeconfig->fileName is set to NULL**
 * 
 * Note: One may use either kubeconfig->fileName or kubeconfig->buffer but not both at the same time.`

File: kubernetes/config/kube_config_yaml.c (hunk 4 above)
Comment (DanyT) [reply kubernetes/config/kube_config_yaml.c:439]: Done.

File: kubernetes/config/kube_config.c (hunk 1 above)
Comment (DanyT) [reply kubernetes/config/kube_config.c]: Meanwhile I added support to use both heap-based and stack-based allocated kubeconfig_t and a bit of documentation for these internal functions.

**Comment by brendandburns** [comment]
This looks good to me. Thanks for the quick turn around. I'll wait for @ityuhui to take a look though in case there are additional comments.

**Comment by ityuhui** [comment]
Thanks @brendandburns 

@DanyT Can you please fix the issues reported by the code-static-check and code-sytle-check https://github.com/kubernetes-client/c/actions/runs/9781006686/job/27005098287?pr=242 in your code changes ?

You can run the checks locally with

```bash
sh ./code-check/code-static-check.sh ./kubernetes/config/
```
and

```bash
find ./kubernetes/config/ -type f -regextype posix-extended -regex ".*\.(c|h)" -exec sh ./code-check/code-style-check.sh {} \;
```

**Comment by ityuhui** [comment]
@DanyT 
Could you please write an example that uses your new function so we can add it to the automated tests in case your functions are broken by future code changes.

File: kubernetes/config/kube_config_yaml.c [hunk 5]
```diff
@@ -430,25 +430,39 @@ int kubeyaml_load_kubeconfig(kubeconfig_t * kubeconfig)
...
     if (kubeconfig->fileName) {
         input = fopen(kubeconfig->fileName, "rb");
         if (!input) {
             fprintf(stderr, "%s: Cannot open the file %s.[%s]\n", fname, kubeconfig->fileName, strerror(errno));
             return -1;
         }
-    } else {
-        fprintf(stderr, "%s: The kubeconf file name needs be set by kubeconfig->fileName .\n", fname);
+    }
+    else if (kubeconfig->buffer) {
```
Comment (ityuhui) [inline kubernetes/config/kube_config_yaml.c:447]: If this condition is not currently in use, delete it.

Comment (DanyT) [reply kubernetes/config/kube_config_yaml.c:447]: This is needed in order to allow code to continue if `kubeconfig->buffer` is set. I did it like this in order to avoid code duplication on parser and document variable initialization

**Comment by DanyT** [comment]
@ityuhui I have added:

- example for load_kube_config_buffer: ./example/list_pod_buffer
- fixed static and style code checks.

Let me know if there is something that i missed.

**Comment by ityuhui** [comment]
@DanyT

It seems that the build failed
```
cd list_pod_buffer; make test
make[1]: Entering directory '/home/runner/work/c/c/examples/list_pod_buffer'
./list_pod_buffer_bin
Cannot get kubernetes configuration from file.
make[1]: *** [Makefile:11: test] Error 255
make[1]: Leaving directory '/home/runner/work/c/c/examples/list_pod_buffer'
make: *** [Makefile:40: test] Error 2
Error: Process completed with exit code 2.
```

File: examples/exec_provider/list_pod_by_exec_provider.c~
Comment (ityuhui) [inline examples/exec_provider/list_pod_by_exec_provider.c~:1]: Remove this file

File: examples/exec_provider/my_exec_provider.c~
Comment (ityuhui) [inline examples/exec_provider/my_exec_provider.c~:1]: Remove this file too.

File: kubernetes/config/kube_config_yaml.c (hunk 5 above)
Comment (ityuhui) [reply kubernetes/config/kube_config_yaml.c]: Got it. Thank you for clarification.

**Comment by ityuhui** [comment]
Nothing special, the configuration is located in `$HOME/.kube/config`

FYI
https://github.com/kubernetes-client/c/blob/dc7d67098361c3579e3dd8153ccc6c4c30ddfedb/kubernetes/config/kube_config.c#L130

**Comment by DanyT** [comment]
@ityuhui  Sorry for the delay. This should fix the issue.

**Comment by ityuhui** [comment]
Thank you @DanyT

Can you add your example to the memory leak test ？Sorry for not reminding you earlier.

https://github.com/kubernetes-client/c/blob/dc7d67098361c3579e3dd8153ccc6c4c30ddfedb/examples/Makefile#L53

**Comment by DanyT** [comment]
Done

**Comment by ityuhui** [comment]
I have no question about this PR. Thanks to @DanyT for the contribution !

@brendandburns I'm OK with the changes.

**Review by ityuhui — APPROVED** [review]
